libs/test-utils/src/expect_print.ts

```diff
@@ -0,0 +1,144 @@
+import { render, RenderResult, waitFor } from '@testing-library/react';
+import {
+  ElementWithCallback,
+  Optional,
+  PrintOptions,
+} from '@votingworks/types';
+import { advancePromises } from './advance_timers';
+import { assert } from './assert';
+
+export class ExpectPrintError extends Error {}
+
+export type InspectPrintFunction = (
+  renderResult: RenderResult,
+  printOptions?: PrintOptions
+) => void;
+
+let lastPrintedElement: Optional<JSX.Element>;
+let lastPrintedElementWithCallback: Optional<ElementWithCallback>;
+let lastPrintOptions: Optional<PrintOptions>;
+
+export function fakePrintElement(
+  element: JSX.Element,
+  printOptions: PrintOptions
+): Promise<void> {
+  if (lastPrintedElement || lastPrintedElementWithCallback) {
+    throw new ExpectPrintError(
+      'You have not made any assertions against the last print before printing another within a test.'
+    );
+  }
+  lastPrintedElement = element;
+  lastPrintOptions = printOptions;
+  return Promise.resolve();
+}
+
+export function fakePrintElementWhenReady(
+  elementWithCallback: ElementWithCallback,
+  printOptions: PrintOptions
+): Promise<void> {
+  if (lastPrintedElement || lastPrintedElementWithCallback) {
+    throw new ExpectPrintError(
+      'You have not made any assertions against the last print before printing another within a test.'
+    );
+  }
+  lastPrintedElementWithCallback = elementWithCallback;
+  lastPrintOptions = printOptions;
+  return Promise.resolve();
+}
+
+/**
+ * Renders the element passed to fakePrintElement and allows assertions
+ * against its content.
+ *
+ * @param inspect method which runs queries against the render result
+ */
+function inspectPrintedElement(inspect: InspectPrintFunction): void {
+  assert(lastPrintedElement);
+  const renderResult = render(lastPrintedElement);
+  inspect(renderResult, lastPrintOptions);
+  renderResult.unmount();
+}
+
+/**
+ * Renders the element with callback passed to fakePrintElementWhenReady
+ * and allows assertions against its content.
+ *
+ * @param inspect method which runs queries against the render result
+ */
+async function inspectPrintedWhenReadyElement(
+  inspect: InspectPrintFunction
+): Promise<void> {
+  let renderResult: Optional<RenderResult>;
+  const renderedPromise = new Promise<void>((resolve) => {
+    assert(lastPrintedElementWithCallback);
+    renderResult = render(lastPrintedElementWithCallback(resolve));
+  });
+  // We have to advance promises here to allow effects to run before
+  // running assertions against the rendered element.
+  await advancePromises();
```

Comment:
When I was first making this, I found that the `advancePromises` was actually necessary in the case of the QR code component, even though I thought at the time it didn't make sense because `onRendered` should have been called. I thought it had something to do with the fact that the effect waited on another async call.

I then refactored that component, and I was probably just doing something wrong to begin with. It doesn't really change anything for now because `inspectPrintedWhenReadyElement` is not currently being used anywhere since HMPB's are currently our only component that need to be printed / expected as such, and will only be relevant to election manager.

I think it makes sense to remove for now and revisit if we have problems using it for the HMPB.

W.r.t testing, we can (and are testing) that the print only happens after the `onRendered` callback. I'm not sure if/what you're suggesting to test in addition. It's on the calling component to be ready to print, assuming the shenanigans you point out here are truly unnecessary.

Reply:
Removed.

Reply:
Cool, thanks for sharing that context. I was suggesting that this `await advancePromises()` might actually give the rendered component _extra_ time to run more effects, which would add a little gap between the `onRendered` callback being called and our assertions (which simulate printing happening). So removing it will better simulate printing immediately.

In summary, I don't think any additional changes are needed other than removing this line.